### DISCUSSION TITLE
Use mitmproxy offline using a custom server certificate

### DISCUSSION THREAD

**Discussion by Zabrane** [mitmproxy/mitmproxy#6470] [General]
Hi MITM experts,

Newbie here.

I followed the instructions explaining how to [use a custom server certificate](https://docs.mitmproxy.org/dev/concepts-certificates/#using-a-custom-server-certificate).

```bash
$ mitmdump  --http2 --listen-host="127.0.0.1" -p 8443 --set upstream_cert=false --certs www.mydomain.com=mydomain.pem
[16:06:15.319] HTTP(S) proxy listening at 127.0.0.1:8443
```

Then, `I turned off the WIFI`.

When my app loads the page `https://www.mydomain.com/`, it seems to me that `mitmdump` tries to connect to live internet in order to build the dummy certificates:
```bash
16:01:18.231][127.0.0.1:59250] client connect
[16:01:18.234][127.0.0.1:59250] error establishing server connection: [Errno 8] nodename nor servname provided, or not known
127.0.0.1:59250: CONNECT www.mydomain.com:443
              << 502 Bad Gateway 227b
[16:01:18.236][127.0.0.1:59250] client disconnect
[16:01:18.246][127.0.0.1:59251] client connect
[16:01:18.250][127.0.0.1:59251] error establishing server connection: [Errno 8] nodename nor servname provided, or not known
```

Finally, I also tried the `--set upstream_cert=false` option without success.

My config:
```
$ mitmdump --version
Mitmproxy: 10.1.3
Python:    3.11.5
OpenSSL:   OpenSSL 3.1.4 24 Oct 2023
Platform:  macOS-13.6-x86_64-i386-64bit
```

Help appreciated.

**Comment by mhils** [comment]
You likely want to set [`connection_strategy`](https://docs.mitmproxy.org/stable/concepts-options/#connection_strategy) to lazy. :)

**Comment by Zabrane** [comment]
@mhils using `--set connection_strategy=lazy` worked as expected (WiFi off) ;-)

I've also noticed that I don't need to setup any custom server dummy certificates.
This works offline as well :
```bash
mitmdump  --http2 --listen-host="127.0.0.1" -p 8443 --set upstream_cert=false
```

Does `mitmdump` stores all (previous) dummy certificates on disk? Is my understanding correct?
If yes, where are there stored?

**Reply by mhils** [reply]
https://docs.mitmproxy.org/stable/concepts-certificates/

Leaf certs are not stored on disk, only the CA is.

**Reply by Zabrane** [reply]
@mhils Thank you. You made my day.
